src/background_music/Integrator.h

```diff
@@ -0,0 +1,222 @@
+     * history multiplier is not valid, std::invalid_argument is thrown.
+     * The choice was made not to return -1 on error, as that easily leads to
+     * segfaults.
+     */
+    [[maybe_unused]] static inline S
+    calculate_samples_from_history_multiplier(S history_multiplier)
+    {
+        double absolute = fabs(history_multiplier);
+        if (absolute < std::numeric_limits<S>::epsilon()())
+        {
+            return 0;
+        }
+        else if (absolute <= (1 - std::numeric_limits<S>::epsilon()()))
+        {
+            return 1.0 / log(absolute);
+        }
+        throw std::invalid_argument(
+            "IntegratorBase::calculate_samples_from_history_multiplier("
+            "multiplier): multiplier out of range.");
+    }
+
+    /**
+     * By default, the integrator does not integrate and just passes-through the
+     * input.
+     */
+    inline constexpr Integrator() : history_multiplier_(0), input_multiplier_(1) {}
+
+    /**
+     * Creates an integrator over the provided number of samples.
+     * @param samples The number of samples
+     */
+    [[maybe_unused]] inline explicit Integrator(S samples)
+        : history_multiplier_(
+              calculate_history_multiplier_from_samples(samples)),
+          input_multiplier_(1.0 - history_multiplier_)
+    {
+    }
+
+    /**
+     * Creates an integrator of the the provided number of samples that is
+     * calculated from the time and sample rate.
+     * @param time The number of time units
+     * @param sample_rate The sample-rate samples per unit of time
+     */
+    [[maybe_unused]] inline Integrator(S time, S sample_rate)
+        : Integrator(time * sample_rate)
+    {
+    }
+    template<typename T1, typename T2>
+    [[maybe_unused]] inline Integrator(T1 time, T2 sample_rate)
+        : Integrator(static_cast<S>(time) * static_cast<S>(sample_rate))
+    {
+        // Replace with requires in on C++20 adoption
+        static_assert(std::is_arithmetic_v<T1>);
+        static_assert(std::is_arithmetic_v<T2>);
+    }
+
+    /**
+     * Integrates value input using history as integrated value.
+     * @param history The historic integrated value
+     * @param input The new input
+     */
+    [[maybe_unused]] inline void integrate(S & history, S input) const
+    {
+        history = history_multiplier_ * history + input_multiplier_ * input;
+    }
+
+    /**
+     * Integrates value input using history as integrated value.
+     * @param history The historic integrated value
+     * @param input The new input
+     */
+    [[maybe_unused]] inline S get_integrated(S history, S input) const
+    {
+        return history_multiplier_ * history + input_multiplier_ * input;
+    }
+
+
+    /**
+     * Getters
+     */
+    [[nodiscard]] [[maybe_unused]] S history_multiplier() const { return history_multiplier_; }
+
+    [[nodiscard]] [[maybe_unused]] S input_multiplier() const { return input_multiplier_; }
```

Comment:
This is a self-contained plugin, not a general-purpose shared library. If there is unused code, please delete it.